### PR TITLE
Implement DNS caching for queries blocked upstream

### PR DESCRIPTION
**Pull request by DL6ER**
# What does this implement/fix?

Implement DNS caching for queries blocked upstream (`NXDOMAIN + no RA bit, `0.0.0.0/::`, and known Umbrella IP blocking pages) accompanied with suitable CI testing. This change improves handling of externally blocked queries in two ways:

1. Once we know the domain is externally blocked, we don't forward it again for the same client (it is currently forwarded each time)
2. We can recognize cache content with the specified addresses/flags as being upstream blocked and don't return them as "OK (cached)" through the API/in the database

---

**Related issue or feature (if applicable):** N/A

**Pull request in [docs](https://github.com/pi-hole/docs) with documentation (if applicable):** N/A

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code.
4. I am willing to help maintain this change if there are issues with it later.
5. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
6. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

## Checklist:

- [x] The code change is tested and works locally.
- [x] I based my code and PRs against the repositories `developmental` branch.
- [x] I [signed off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits. Pi-hole enforces the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions
- [x] I [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) all my commits. Pi-hole requires signatures to verify authorship
- [x] I have read the above and my PR is ready for review.

### CODE REVIEW

**Comment by Anuskuss** [comment]
I set up my upstream to respond with `NXDOMAIN` + no `ra` (+ [EDE15](https://www.rfc-editor.org/rfc/rfc8914#name-extended-dns-error-code-15-)) and this works for me.
```
$ dig @127.0.0.1 -p53 +noall +answer +comments one.one.one.one 
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 56340
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
; EDE: 15 (Blocked): (1.1.1.1)
;; ANSWER SECTION:
one.one.one.one. 2	IN	A	0.0.0.0
```
```
$ dig @127.0.0.1 -p53 +noall +answer +comments one.one.one.one
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 27011
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; ANSWER SECTION:
one.one.one.one. 2	IN	A	0.0.0.0
```

Unfortunately the EDE message doesn't get cached yet (notice how it disappears in the second request) but this is way better than before. Now it actually responds with my `BLOCKINGMODE` (`NULL` in my case) and shows `Blocked (external, NXRA)` both times in the log instead of `OK (cache)` the second time. Good job :+1:

**Comment by DL6ER** [comment]
We have a follow-up already in the pipeline (branch `new/edns_reasoning`) but I will open the PR only after this has been merged to ease reviewing. It will offer EDE to the clients why the Pi-hole has decided to block queries. This feature can be controlled by a new setting `dns.blocking.edns`

Examples:
``` bash
$ dig doubleclick.net | grep EDE
; EDE: 29: (synthesized)

$ dig doubleclick.net | grep EDE
; EDE: 15 (Blocked): (gravity)

$ dig denied.ftl | grep EDE
; EDE: 15 (Blocked): (denylist)

$ dig some.regex.example.com | grep EDE
; EDE: 15 (Blocked): (regex)

$ dig null.ftl | grep EDE
; EDE: 15 (Blocked): (upstream NULL)
```

**Review by yubiuser — COMMENTED** [review]
> Once we know the domain is externally blocked, we don't forward it again for the same client 

Not forwarding for how long? After a certain time we should allow forwarding again - the domain might not be blocked anymore

**Comment by DL6ER** [comment]
> Not forwarding for how long?

This is a good question - something I have not thought about. So far, the cache is kept valid until something changes *locally* (gravity update, changes to the domainlists, etc.). But with this PR we may need to make DNS records actually expire... I wonder: is "it is cleared on the next gravity run" too long?

File: src/enums.h [hunk 1]
```diff
@@ -130,6 +130,9 @@ enum domain_client_status {
 	REGEX_BLOCKED,
 	ALLOWED,
 	SPECIAL_DOMAIN,
+	UPSTREAM_BLOCKED_NXRA,
```
[the remark this answers is not in the document]
Comment (DL6ER) [reply src/enums.h:133]: No, they are only internally used meta-states. The cache is never exported to the database.

**Comment by yubiuser** [comment]
>  I wonder: is "it is cleared on the next gravity run" too long?

Personally, I think yes. Such a feature reminds me to `unbound`'s `serve-expired` with a default limit of 1 day. 
https://unbound.docs.nlnetlabs.nl/en/latest/manpages/unbound.conf.html#unbound-conf-serve-expired-ttl

**Comment by rdwebdesign** [comment]
>  I wonder: is "it is cleared on the next gravity run" too long?

This raises another question: Will we add a new config option to change this expiration time?

I agree with yubiuser. `1 day` seems to be a good default value.

**Comment by Anuskuss** [comment]
Personally I don't think that's necessary. I think that if upstream decides to outright block a domain it has do be something serious (e.g. illegal). So I think once a week is plenty.

**Comment by DL6ER** [comment]
New config option added, defaulting to one day - as suggested. Note that `pihole -g` will still flush the entire cache *including* the queries blocked upstream.

**Comment by DL6ER** [comment]
A commit concerning EDE sneaked into this PR which I wanted to keep separate to ease reviewing. Force-pushed away.

File: src/api/docs/content/specs/config.yaml [hunk 2]
```diff
@@ -239,6 +239,8 @@ components:
                       type: integer
                     optimizer:
                       type: integer
+                    upstreamTTL:
```
Comment (yubiuser) [inline src/api/docs/content/specs/config.yaml:242]:
```suggestion
                    upstreamBlockedTTL:
```

**Review by yubiuser — COMMENTED** [review]
Can you squeeze in `Blocked` in the config option name? Just to make it clear that this option does not 'overwrite' all TTLs form upstream responses, but just set them for blocked queries.

File: src/api/docs/content/specs/config.yaml (hunk 2 above)
Comment (DL6ER) [reply src/api/docs/content/specs/config.yaml]: done in the latest commit, needed to be done in 12 locations

**Review by yubiuser — APPROVED** [review]

**Pull request merged**